Extending `ISetupEventHandler` with new events

MikeAlhayek opened the pull request:
Fix #14170

@jtkech we discussed this in #14172.

This is the approach I had in mind. With the new `ITenantSetupHandler`, one can perform some setup logic after the tenant is setup.

Thoughts?

MikeAlhayek commented:
@jtkech I am noticing an issue here that I can't seem to explain. 

```C#
public class SelfTenantSetupHandler : TenantSetupHandlerBase
{
    private readonly INotifier _notifier;
    private readonly IHtmlLocalizer H;

    public TestSetup(
        INotifier notifier,
        IHtmlLocalizer<TestSetup> localizer)
    {
        _notifier = notifier;
        H = localizer;
    }

    public override async Task SuccessAsync()
    {
        await _notifier.SuccessAsync(H["You successfully have setup your website!"]);
    }
}
```

When the tenant is done setting up, we set set it's status to running and then reload the site. So `SuccessAsync()` is called in a new scope "after the site is reloaded". I am expecting that the above `SelfTenantSetupHandler` to show show the notification message. But it does not! Why would it not show the notification here? It's like the message is added to the current HttpContext which gets lost when we start a new scope?

jtkech commented:
Yes `INotifier` is a scoped service so what you enlist in a new local inner scope is lost when returning back to the controller scope.

---

Yes I closed my PR to make a pause around this, waiting for after the incoming release and because I have some works in late, but I'm not against so let's restart the discussion ;)

Maybe we could have kept the existing `ISetupEventHandler` and just extended it, yes by making a `SetupEventHandlerBase` class so that custom implementations would have just to inherit from `SetupEventHandlerBase` in place of `ISetupEventHandler` to migrate.

I like the method names, if we use `Failed()` we could then use `Succeeded()`, hmm I think we could keep the "Setup" term, so `SetupAsync()` in place of `SettingUpAsync()`, anyway at this point the setup is almost done, only handlers may add errors. And maybe in this order.

    Task SetupAsync(SetupContext context);
    Task FailedAsync(SetupContext context);
    Task SucceededAsync();

Just saw that if we failed to init the data store we alreay have enlisted an error, I think we are missing the following check also before reloading the shell.

            if (context.Errors.Any())
            {
                // Here do we need to call the failed event?
                return null;
            }

Or maybe handlers would have to check if there is already an error if we don't add the above check, but yes existing handlers don't have a direct access to the setup context, maybe there is a good reason.

Not fully sure we need a `Faulted()` event, what can we expect to do consistently if the tenant is not setup correctly, maybe here we are not at the right level to manage failed setups, we may need a host level service and / or handlers to manage failed and successful setups, but not fully sure ;)

I'm stopping, I will think about it and review it more in depth when I will have time.

MikeAlhayek commented:
> Maybe we could have kept the existing ISetupEventHandler and just extended it, yes by making a SetupEventHandlerBase class so that custom implementations would have just to inherit from SetupEventHandlerBase in place of ISetupEventHandler to migrate.

That would be a breaking change which is I am trying to avoid. Other people may already be implementing `ISetupEventHandler` directly in their project.

Not sure I follow you on the `FailedAsync()`. We currently have a way to report error which cancels the setup. But we don't have a way of knowing if the setup failed. Checking the errors does in the handler will not give us accurate status since the error could be called on the next handler. 

I'll rename events as per your suggestions. Also check the condition that returns null.

jtkech commented:
> That would be a breaking change which is I am trying to avoid. Other people may already be implementing ISetupEventHandler directly in their project.

Yes I know that's why in my other PR (whose goal was not the same) I let it as is and only added a new handler. Here I thought about what would be better at the end, sometimes we can introduce intentionally a well documented breaking change, but as said I was not fully sure, and yes here maybe not worth to do.

Hmm, maybe a compromise would be to keep `ISetupEventHandler` as it is, this one being to collaborate to the setup itself, only done locally, then add a new interface, whatever the name, but to only handle the 2 new tenant setup events, the success event and now the failed event that you added.

Otherwise why not, maybe I'm wrong, only thinkings on the fly, as said would need to focus on it a little more.

You already described a little what you may do in a setup succeeded event, can you describe what you may do in a setup failed event?

MikeAlhayek commented:
@jtkech `FailedAsync` could be used for logging or communication. For example, create support ticket in a help desk to alert that a self setup is failing the customer may need help. Or maybe one want to send an email to the SaaS support.

I went ahead and did some cleanup in the `SetupService`. I move shell state alteration out of the `SetupInternalAsync` so it is easier to understand the tenant status changes in one place. Also, I moved reporting success or error to the `SetupAsync` this way the code is much cleaner and we remove redundant code.

Let me know if you see any issues with the presented PR.

jtkech commented:
Okay, 

Looks like you did too much changes and are doing things in some places that may not be correct.

I will review and comment it asap.

jtkech commented on src/OrchardCore/OrchardCore.Setup.Core/SetupService.cs:
```diff
@@ -89,54 +89,84 @@ public async Task<IEnumerable<RecipeDescriptor>> GetSetupRecipesAsync()
                 {
                     context.ShellSettings.State = initialState;
-                    await _shellHost.ReloadShellContextAsync(context.ShellSettings, eventSource: false);
+                    await _shellHost.ReloadShellContextAsync(shellSettings, eventSource: false);
```
This is wrong, here to reload the shell we should use `context.ShellSettings` on which we revert back the state to the initial state, this is important to keep the tenant state consistent. Note: The reload may retrieve an updated state if for example the tenant was setup by another instance, which may explain why it failed here on this instance.

We should not use the internal `shellSettings` that we created locally only in memory for the setup, here if there is an error it would be still in the `Initializing` state, so the reload will not work as expected. We can only use this `ShellSettings` after a successful setup including the call to `UpdateShellSettingsAsync()`, as you do below for the `Succeeded` event.

MikeAlhayek replied:
@jtkech what is a bit confusing for me here is why would be then call `_shellHost.UpdateShellSettingsAsync(shellSettings.AsRunning())` on the shellSettings that is in memory but then never set `context.ShellSettings.AsRunning()`. Does this mean when we call `UpdateShellSettingsAsync` on the memory ShellSettings, it'll override by writing it to the ShellSettings files which makes `context.ShellSettings` irrelevant?

jtkech commented on src/OrchardCore/OrchardCore.Setup.Core/SetupService.cs:
```diff
@@ -89,54 +89,84 @@ public async Task<IEnumerable<RecipeDescriptor>> GetSetupRecipesAsync()
+                    {
+                        var tenandSetupHandlers = scope.ServiceProvider.GetServices<ITenantSetupHandler>();
+
+                        await tenandSetupHandlers.InvokeAsync((handler) => handler.FailedAsync(context), _logger);
```
Okay when at least you were calling these handlers below after the existing setup handlers that may add errors, this because the setup was almost done.

But here we also may just fail or retrieve an `Uninitialized` state without any features (of the setup recipe) enabled and then no tenant setup handlers to be resolved.

That's why for the `Failed` event I was thinking about a host level service, or at least a `DefaultTenantOnly` service belonging to a feature that would need to be part of the `Setup Features` , this to not let the tenant itself to do something if it failed to be setup.

But as said maybe okay if done after the existing setup handlers as you did previously.

jtkech commented on src/OrchardCore/OrchardCore.Setup.Core/SetupService.cs:
```diff
@@ -89,54 +89,84 @@ public async Task<IEnumerable<RecipeDescriptor>> GetSetupRecipesAsync()
+
                 context.ShellSettings.State = initialState;
-                await _shellHost.ReloadShellContextAsync(context.ShellSettings, eventSource: false);
+                await _shellHost.ReloadShellContextAsync(shellSettings ?? context.ShellSettings, eventSource: false);
```
Same comment

MikeAlhayek replied:
Yes I think I understand the problem you pointed out here.

jtkech commented on src/OrchardCore/OrchardCore.Setup.Core/SetupService.cs:
```diff
@@ -89,54 +89,84 @@ public async Task<IEnumerable<RecipeDescriptor>> GetSetupRecipesAsync()
+                {
+                    var tenandSetupHandlers = scope.ServiceProvider.GetServices<ITenantSetupHandler>();
+
+                    await tenandSetupHandlers.InvokeAsync((handler) => handler.FailedAsync(context), _logger);
```
Same comments

MikeAlhayek replied:
done

jtkech commented on src/OrchardCore/OrchardCore.Setup.Core/SetupService.cs:
```diff
@@ -89,54 +89,84 @@ public async Task<IEnumerable<RecipeDescriptor>> GetSetupRecipesAsync()
+                {
+                    var tenandSetupHandlers = scope.ServiceProvider.GetServices<ITenantSetupHandler>();
+
+                    await tenandSetupHandlers.InvokeAsync((handler) => handler.SucceededAsync(), _logger);
```
That's okay for the succeeded event, here or as we did previously.

If we only have this new event, we could just add this as I suggested but by not using a background job.

If we keep the `Failed` event and if it should be managed differently (not by the related tenant itself), maybe better to do the same for the `Suceeded` event too.

MikeAlhayek replied:
I moved the `FailedAsync` call just after we call `SetupAsync()` so we can report the error in the same scope. Let me know if you see other issues here.

jtkech commented on src/OrchardCore/OrchardCore.Setup.Core/SetupService.cs:
```diff
@@ -89,54 +89,84 @@ public async Task<IEnumerable<RecipeDescriptor>> GetSetupRecipesAsync()
             // Due to database collation we normalize the userId to lower invariant.
             // During setup there are no users so we do not need to check unicity.
-            context.Properties[SetupConstants.AdminUserId] = _setupUserIdGenerator.GenerateUniqueId().ToLowerInvariant();
+            context.Properties[SetupConstants.AdminUserId] = _setupUserIdGenerator.GenerateUniqueId();
```
Hmm, maybe possible that we resolve a generator that doesn't only provide lower cases.

jtkech replied:
Because of this comment

            // Due to database collation we normalize the userId to lower invariant.

MikeAlhayek replied:
yes but wouldn't that be a problem everywhere since we use the same `GenerateUniqueId()` everywhere? If the collation can't support uppercases, then either the collation should be changed or the `GenerateUniqueId()`. changing it here only seems very bad to me.

jtkech commented on src/OrchardCore/OrchardCore.Setup.Core/SetupService.cs:
```diff
@@ -89,54 +89,84 @@ public async Task<IEnumerable<RecipeDescriptor>> GetSetupRecipesAsync()
+                });
 
-                throw;
+                return null;
```
Not sure I agree, in this catch block I would still only reload the shell and then re-throw.

MikeAlhayek replied:
Are you sure about this? I am looking at how we call `SetupAsync() from `AutoSetupMiddleware` or `SetupController` and we are not catching that error if an exception was thrown. With the change, we log the error into the `Errors` collection and we prevent a system exception.

MikeAlhayek replied:
@jtkech after reading my last comment, do you still think we should `throw` or `return executionId` is enough?

MikeAlhayek commented:
@jtkech I reverted some changes here and tested it. Everything seems to be still working as expected. I'll await your reply to my comments before making the other changes. I also changed where we fire the new events based on your recommendation.

jtkech commented:
Okay, can't right know but will review asap

jtkech commented on src/OrchardCore.Modules/OrchardCore.Settings/Services/SetupEventHandler.cs:
```diff
@@ -1,15 +1,15 @@
     /// During setup, registers the Super User.
     /// </summary>
-    public class SetupEventHandler : ISetupEventHandler
+    public class SetupEventHandler : TenantSetupHandlerBase
```
Should we then change the class name to `TenantSetupHandler`.

The problem is that here I prefer the old interface name ending with `EventHandler`.

So maybe `ITenantSetupEventHandler` and `TenantSetupEventHandlerBase` but looks too long.

Maybe worth that you ask to @sebastienros in the next triage if we could introduce a little breaking change as I suggested by keeping the `ISetupEventHandler` and just extending it with the new events with the help of a `SetupEventHandlerBase` class. And maybe also provide an extension marked as obsolete to be still able to call `Setup()` in place of `SetupAsync()`.

So that we have a clean interface by keeping the right name, no need to use a pragma and so on.

MikeAlhayek replied:
I'll ask. But, with the current approach we do not need to introduce the breaking change.

jtkech replied:
Yes, just to confirm because ATM I prefer the name of the old interface.

MikeAlhayek replied:
@sebastienros what do you think here? Can we introduce a small breaking change? In summary we want to modify `ISetupEventHandler` interface from this

```C#
public interface ISetupEventHandler
{
    Task Setup(IDictionary<string, object> properties, Action<string, string> reportError);
}
```

to this

```C#
public interface ISetupEventHandler
{
    Task SetupAsync(SetupContext context);

    Task FailedAsync(SetupContext context);

    Task SucceededAsync();
}
```

This will break anyone that is using `ISetupEventHandler` in their project. This should be a low impact breaking change since `ISetupEventHandler` would not be widely used.

jtkech replied:
Yes, I would go this way, much cleaner, but worth to wait for a confirmation.

If no feedback, let's still go this way ;)

MikeAlhayek replied:
@jtkech Learned something new today! As it strange as it may sounds, but starting with C# 8.0, interfaces can have default implementation! Not a very good approach and we should avoiding doing this. But, it this case it make sense to use it to prevent breaking change. 

Check out the new changes. But I think it's pretty clean now.

jtkech replied:
Okay cool then, I will look at it asap.

jtkech commented on src/OrchardCore/OrchardCore.Setup.Abstractions/Events/ITenantSetupHandler.cs:
```diff
@@ -0,0 +1,26 @@
+    /// Called when a tenant fails to setup.
+    /// </summary>
+    /// <returns></returns>
+    Task FailedAsync(SetupContext context);
```
I would define these methods in the same order than in the base class.

And in this order `SetupAsync()`, `FailedAsync()` and `SucceededAsync()`.

jtkech commented on src/OrchardCore/OrchardCore.Setup.Abstractions/TenantSetupHandlerBase.cs:
```diff
@@ -0,0 +1,17 @@
+        => Task.CompletedTask;
+
+    public virtual Task SucceededAsync()
+        => Task.CompletedTask;
```
Maybe in one line but as you want.

But in the same order than in the interface, see the related comment.

jtkech commented on src/OrchardCore/OrchardCore.Setup.Core/SetupService.cs:
```diff
@@ -189,71 +189,89 @@ private async Task<string> SetupInternalAsync(SetupContext context)
             await _shellHost.UpdateShellSettingsAsync(shellSettings.AsRunning());
 
+            var runningShellScope = await _shellHost.GetScopeAsync(shellSettings);
+            await runningShellScope.UsingAsync(async scope =>
```
I would do it in one line, as we do elsewhere.

MikeAlhayek commented on src/OrchardCore/OrchardCore.Setup.Core/SetupService.cs:
```diff
@@ -189,71 +189,89 @@ private async Task<string> SetupInternalAsync(SetupContext context)
-                if (context.Errors.Any())
+                if (context.Errors.Count == 0)
                 {
-                    return null;
```
The return is useless. The next check will always take place as it should

jtkech commented on src/OrchardCore/OrchardCore.Setup.Core/SetupService.cs:
```diff
@@ -92,57 +92,57 @@ public async Task<string> SetupAsync(SetupContext context)
-            // Features to enable for Setup
-            string[] hardcoded =
+            // Features to enable for Setup.
+            string[] coreFeature =
```
`coreFeature` => `coreFeatures`

MikeAlhayek commented:
@jtkech I think I addressed all you feedback. Let me know if I missed anything.

MikeAlhayek commented on src/OrchardCore/OrchardCore.Setup.Core/SetupService.cs:
```diff
@@ -189,71 +189,89 @@ private async Task<string> SetupInternalAsync(SetupContext context)
+
+#pragma warning disable CS0618 // Type or member is obsolete
+                var setupEventHandlers = scope.ServiceProvider.GetServices<ISetupEventHandler>();
+#pragma warning restore CS0618 // Type or member is obsolete
```
Added the `, false` but we still have to add the `#pragma` to avoid the warning.

jtkech commented:
Okay, looks good, will review tomorrow.

jtkech commented on src/OrchardCore/OrchardCore.Setup.Core/SetupService.cs:
```diff
@@ -89,60 +89,57 @@ public async Task<IEnumerable<RecipeDescriptor>> GetSetupRecipesAsync()
-                var executionId = await SetupInternalAsync(context);
+                // Set the shell state to 'Initializing' to guarantee that any subsequent HTTP requests
+                // will result in a 'Service Unavailable' response while the tenant setup is in progress.
+                context.ShellSettings.AsInitializing();
```
I suggest to keep the code of all this method as it was before.

MikeAlhayek replied:
It's very weird/not-readable to change the status into 2 methods. What I made here will altering the state of the same object in the same method.

jtkech replied:
This method reverts the state without having to know which state will be set in the other method. Yes maybe this separation is not needed in the same class, I don't know why it was done like this, maybe these 2 methods were initially in 2 separate services.

 I already thought about it but I just kept the code unchanged. Otherwise we could also say that we can only setup an `Uninitialized` tenant and then the `initialState` becomes useless. So I'm not against but in a separate PR with maybe more cleanups.

jtkech commented on src/OrchardCore/OrchardCore.Setup.Core/SetupService.cs:
```diff
@@ -89,60 +89,57 @@ public async Task<IEnumerable<RecipeDescriptor>> GetSetupRecipesAsync()
                 context.ShellSettings.State = initialState;
                 await _shellHost.ReloadShellContextAsync(context.ShellSettings, eventSource: false);
-
-                throw;
```
Before there was no catch at all.

I added this catch to at least revert the shell state and reload it to keep it in a consistent state, before rethrowing.

I suggest to keep this code as it was.

Maybe in a separate PR if really better to return back to the controller.

jtkech commented on src/OrchardCore/OrchardCore.Setup.Core/SetupService.cs:
```diff
@@ -89,60 +89,57 @@ public async Task<IEnumerable<RecipeDescriptor>> GetSetupRecipesAsync()
-            string executionId;
+            _logger.LogInformation("Running setup for tenant '{TenantName}'.", context.ShellSettings?.Name);
 
-            if (_logger.IsEnabled(LogLevel.Information))
```
Why removing this check?

MikeAlhayek replied:
Because LogInformation will only print to the file when the log level is info+. That check isn't needed

jtkech replied:
But it will still parse and tokenise the string to insert serialized params.

Here not so important but that's a good practice.

MikeAlhayek replied:
If this check is needed it should be part of the logger implementation not our code. Anyway, I'll restore it.

jtkech replied:
It was suggested but, as I remember, at some point they didn't because too late, so to prevent double checking. Maybe in the meantime it has been added, here some interesting links.

https://andrewlock.net/exploring-dotnet-6-part-8-improving-logging-performance-with-source-generators/

https://github.com/dotnet/runtime/issues/63907

https://github.com/dotnet/runtime/issues/51927

jtkech commented on src/OrchardCore/OrchardCore.Setup.Core/SetupService.cs:
```diff
@@ -189,71 +189,89 @@ private async Task<string> SetupInternalAsync(SetupContext context)
+            if (context.Errors.Count > 0)
+            {
+                return null;
             }
```
As a reminder, this can be removed.

As said, until there, unless minor changes, almost all can be kept as before.

jtkech commented on src/OrchardCore/OrchardCore.Setup.Core/SetupService.cs:
```diff
@@ -228,31 +222,49 @@ await scope
-            if (context.Errors.Any())
+            if (context.Errors.Count == 0)
             {
-                return executionId;
```
I like the previous pattern, first check the opposite and return if true, as aspnetcore often do (but not always), it's more readable and prevent indentation afterwards.

But yes here not so important, just to keep the code as it was before as much as possible.

jtkech commented on src/OrchardCore/OrchardCore.Setup.Core/SetupService.cs:
```diff
@@ -89,60 +89,57 @@ public async Task<IEnumerable<RecipeDescriptor>> GetSetupRecipesAsync()
+
             if (context.Properties.TryGetValue(SetupConstants.AdminUsername, out var adminUsername))
             {
                 recipeEnvironmentFeature.Properties[SetupConstants.AdminUsername] = adminUsername;
```
You can add an empty line before the below if block line 147

sebastienros commented:
@MikeAlhayek will test default interfance implementations

jtkech commented on src/OrchardCore/OrchardCore.Setup.Abstractions/Events/ISetupEventHandler.cs:
```diff
@@ -1,17 +1,31 @@
+    /// Called when a tenant fails to setup.
+    /// </summary>
+    /// <returns></returns>
+    Task FailedAsync(SetupContext context) => Task.CompletedTask;
```
Yes, didn't think about this and the default implementation is so simple !!!

jtkech commented on src/OrchardCore/OrchardCore.Setup.Abstractions/SetupEventHandler.cs:
```diff
@@ -0,0 +1,22 @@
+
+namespace OrchardCore.Setup;
+
+public class SetupEventHandler : ISetupEventHandler
```
Maybe we no longer need this base class

jtkech commented on src/OrchardCore.Modules/OrchardCore.Settings/Services/GeneralSetupEventHandler.cs:
```diff
@@ -0,0 +1,34 @@
+    /// <summary>
+    /// During setup, registers the Super User.
+    /// </summary>
+    public class GeneralSetupEventHandler : SetupEventHandler
```
Maybe `SettingsSetupEventHandler`

MikeAlhayek commented on src/OrchardCore/OrchardCore.Setup.Abstractions/Events/ISetupEventHandler.cs:
```diff
@@ -1,17 +1,31 @@
+public interface ISetupEventHandler
 {
+    [Obsolete($"This method is obsolete and will be removed in future releases. Please use '{nameof(SetupAsync)}' instead.", false)]
+    Task Setup(IDictionary<string, object> properties, Action<string, string> reportError);
```
@jtkech 
I was hoping to leave the base class so in the future we can remove the default implementation from the interface. + Today if someone is using the interface they would have to implement the Setup methods.

jtkech replied:
But not if a default implementation is also provided for this remaining one

jtkech replied:
But okay because these default implementations are intended to be removed.

jtkech commented on src/OrchardCore/OrchardCore.Setup.Core/SetupService.cs:
```diff
@@ -92,18 +92,25 @@ public async Task<string> SetupAsync(SetupContext context)
+            catch (Exception e)
             {
+                _logger.LogError(e, "An error occurred while setting up the tenant.");
+                context.Errors.Add(String.Empty, S["An error occurred while setting up the tenants: {0}", e.Message]);
```
If we throw again, we don't need these 2 lines
We don't need to add an error to the setup context
And on throwing, the exception will be logged.

jtkech commented on src/OrchardCore/OrchardCore.Setup.Core/SetupService.cs:
```diff
@@ -92,18 +92,25 @@ public async Task<string> SetupAsync(SetupContext context)
             try
             {
+                // Set the shell state to 'Initializing' to guarantee that any subsequent HTTP requests
+                // will result in a 'Service Unavailable' response while the tenant setup is in progress.
```
Still not a fan to having moved this part (including the below line).

MikeAlhayek commented:
@jtkech Last changes? hopefully...

jtkech commented:
Yes, sorry for being too strict sometimes, maybe because it's too hot in France ;)

jtkech commented on src/OrchardCore/OrchardCore.Setup.Abstractions/Events/ISetupEventHandler.cs:
```diff
@@ -1,17 +1,35 @@
+    /// <param name="context"></param>
+    /// <returns></returns>
+    Task SetupAsync(SetupContext context)
+        => Task.CompletedTask;
```
In the meantime I will approve

But as a last idea the default implementation of `SetupAsync()` could be to call `Setup()` so that we don't need to call the obsolete `Setup()` from the `SetupService` and only use the rule pragma here, all in one place.

Something like this, quickly written so may need to be tweaked.

    Task SetupAsync(SetupContext context) =>
        Setup(context.Properties, (key, message) => context.Errors[key] = message );
